### PR TITLE
kubelet should retry fetching ECR auth token

### LINKED ISSUE #37962: Kubelet doesn't retry if failed to get ECR auth token
---
AWS has a rate limit of 1 ECR auth token per second per account. For kubernetes cluster with many nodes and using ECR as its primary registry its not unlikely for the kubelet to get throttled. 
```
aws_credentials.go:199] while requesting ECR authorization token ThrottlingException: Rate exceeded
```
After digging through some of the code in the [aws_credentials.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/aws/aws_credentials.go)  (and please correct me if I'm wrong) it seems like the kubelet does not implement any retry logic for getting the ECR token which leaves that worker node completely useless until the expiration time for the token (11 hours and 55 minutes). The only way to trigger another fetch is to restart the kublet. The kubelet should either block until the token is acquired OR execute some retry loop in a separate goroutine. 

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.4", GitCommit:"3b417cc4ccd1b8f38ff9ec96bb50a81ca0ea9d56", GitTreeState:"clean", BuildDate:"2016-10-21T02:48:38Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.4", GitCommit:"3b417cc4ccd1b8f38ff9ec96bb50a81ca0ea9d56", GitTreeState:"clean", BuildDate:"2016-10-21T02:42:39Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): 
```
NAME=CoreOS
ID=coreos
VERSION=1185.2.0
VERSION_ID=1185.2.0
BUILD_ID=2016-10-20-2326
PRETTY_NAME="CoreOS 1185.2.0 (MoreOS)"
ANSI_COLOR="1;32"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://github.com/coreos/bugs/issues"
```
- **Kernel** (e.g. `uname -a`):
```
Linux 4.7.3-coreos-r2 #1 SMP Thu Oct 20 23:18:54 UTC 2016 x86_64 Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz GenuineIntel GNU/Linux
```

### PR DESCRIPTION
**Pull request by lukasz-ochedzan-ef**
rebase of @andrewsykim PR that got closed due to inactivity

> /kind feature
Fixes #37962
```release-note
NONE
```

### CODE REVIEW

File: pkg/credentialprovider/aws/aws_credentials.go [hunk 1]
```diff
@@ -36,6 +36,28 @@ const awsStandardDNSSuffix = "amazonaws.com"
 const awsChinaDNSSuffix = "amazonaws.com.cn"
 const registryURLTemplate = "*.dkr.ecr.%s.%s"
 
+// AWSRegions is the complete list of regions known to the AWS cloudprovider
+// and credentialprovider.
+var AWSRegions = [...]string{
```
Comment (andrewsykim) [inline pkg/credentialprovider/aws/aws_credentials.go]: There's PR open to consolidate some of the AWS regions logic https://github.com/kubernetes/kubernetes/pull/70675. I recommend waiting for that PR to go in first :)

Comment (lukasz-ochedzan-ef) [reply pkg/credentialprovider/aws/aws_credentials.go]: :+1:

File: pkg/credentialprovider/aws/aws_credentials.go [hunk 2]
```diff
@@ -50,6 +72,26 @@ func awsHandlerLogger(req *request.Request) {
 	glog.V(3).Infof("AWS request: %s:%s in %s", service, name, *region)
 }
 
+// Implements the RequestRetryer interface
```
Comment (krmayankk) [inline pkg/credentialprovider/aws/aws_credentials.go]: I am guessing the backoff logic is in the RequestRetryer impl somewhere ?

File: pkg/credentialprovider/aws/aws_credentials.go [hunk 3]
```diff
@@ -36,6 +36,28 @@ const awsStandardDNSSuffix = "amazonaws.com"
...
+	"eu-west-1",
+	"eu-central-1",
+	"ap-south-1",
+	"ap-southeast-1",
+	"ap-southeast-2",
+	"ap-northeast-1",
+	"ap-northeast-2",
+	"cn-north-1",
+	"us-gov-west-1",
+	"sa-east-1",
```
Comment (krmayankk) [inline pkg/credentialprovider/aws/aws_credentials.go]: is there an AWS api to get the regions ?

Comment (andrewsykim) [reply pkg/credentialprovider/aws/aws_credentials.go]: There is and was added in https://github.com/kubernetes/kubernetes/pull/70675 :)

**Comment by andrewsykim** [comment]
/assign @mcrute

Please follow-up on this PR when the AWS credential provider improvements are in.

**Comment by andrewsykim** [comment]
/close 

Fixed in https://github.com/kubernetes/kubernetes/pull/75587

**Pull request closed without merging**